test/shared/formatSqrtRatioX96.ts

```diff
@@ -1,21 +1,33 @@
 import { BigNumber } from 'ethers'
 import Decimal from 'decimal.js'
 
+const TWO = BigNumber.from(2)
 const TEN = BigNumber.from(10)
 const FIVE_SIG_FIGS_POW = new Decimal(10).pow(5)
 
-export default function formatSqrtRatioX96(
+export function formatSqrtRatioX96(
   sqrtRatioX96: BigNumber,
   decimalsToken0: number = 18,
   decimalsToken1: number = 18
 ): string {
-  Decimal.set({ precision: 5, toExpPos: 9_999_999, toExpNeg: -9_999_999, rounding: Decimal.ROUND_HALF_CEIL })
+  Decimal.set({ precision: 5, toExpPos: 9_999_999, toExpNeg: -9_999_999, rounding: Decimal.ROUND_HALF_EVEN })
 
-  const decRatio = new Decimal(sqrtRatioX96.toString())
-
-  let ratio = decRatio.div(new Decimal(2).pow(96)).pow(2)
+  let ratio
+  let ratioNum
+  let ratioBN = sqrtRatioX96.mul(sqrtRatioX96).div(TWO.pow(64))
```

Comment:
ideally you just use another mostly self contained implementation that we should expect to be correct, rather than reimplement the solidity logic in js, but as long as we know this implementation is correct it's nbd. 

i think the reason it was off before could be that precision is set to 5 on line 12, but i think i did that for formatting the text. it might be possible to get Decimal to be accurate with higher precision though

Reply:
Even without precision, the Decimals library calculations are not as accurate as BigNumber or even JS number and will have intermittent errors with the full number being slightly off.

The way it is currently now, there are 0 discrepancies, except when doing token decimals adjustments which once again, do very large multiplications and divisions which impact the final result by a bit. The only way to get the same exact result is to do those same mul/divs in javascript. Which honestly kind of sucks..... 

I think we know the implementation is correct now, but it WASN'T. And I knew that thanks to not copying every little solidity thing...